File: backend/py_template/devdonalds.py
from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
import re

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
	name: str

@dataclass
class RequiredItem():
	name: str
	quantity: int

@dataclass
class Recipe(CookbookEntry):
	required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
	cook_time: int


# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook = {}

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
	data = request.get_json()
	recipe_name = data.get('input', '')
	parsed_name = parse_handwriting(recipe_name)
	if parsed_name is None:
		return 'Invalid recipe name', 400
	return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that 
def parse_handwriting(recipeName: str) -> Union[str | None]:
	recipeName = recipeName.replace("_", " ").replace("-", " ")
	recipeName = "".join(char for char in recipeName if char.isalpha() or char.isspace())
	words = recipeName.strip().split()

	if not words:
		return None
	
	return " ".join(word.capitalize() for word in words)

# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def create_entry():
	# Create and add more entries to cookbook
	# Implement endpoint and store the cookbook entries. if successful, return HTTP 200 and empty response body

	data = request.get_json()

	type = data.get('type')
	name = data.get('name')

	if name in cookbook:
		return 'Error: entry already exists! (entry names must be unique)', 400
	
	if type == 'recipe':
		visited_items = set() # set as by storing strings instead of obj, that way can check for duplicate strings
		approved_items = [] # list as final required_items is type List

		items = data.get('requiredItems', [])

		for item in items:
			item_name = item.get('name')
			item_quantity = item.get('quantity')

			if item_name in visited_items:
				return 'Error: duplicated required item in recipe!', 400
			
			visited_items.add(item_name)
			approved_items.append(RequiredItem(name = item_name, quantity = item_quantity))
		
		new_entry = Recipe(name = name, required_items = approved_items)
	elif type == 'ingredient':
		cook_time = data.get('cookTime')
		
		if cook_time < 0:
			return 'Error: invalid cook time! (cook time >= 0)', 400
		
		new_entry = Ingredient(name = name, cook_time = cook_time)
	else:
		return 'Error: invalid type! (not recipe or ingredient)', 400
	
	cookbook[name] = new_entry
	
	return jsonify({}), 200

# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name
@app.route('/summary', methods=['GET'])
def summary():
	name = request.args.get('name')
	if name not in cookbook or not isinstance(cookbook[name], Recipe):
		return 'Error: name not found or is not a recipe!', 400
	
	# Recursion outer variables
	total_cook_time = 0
	final_ingredients_list = {}

	def resolve_item(name, frequency):
		nonlocal total_cook_time
		if name not in cookbook:
			raise ValueError("Missing item in cookbook") # stops recursion
		
		entry = cookbook[name]

		if isinstance(entry, Ingredient):
			total_cook_time += entry.cook_time * frequency
			final_ingredients_list[name] = final_ingredients_list.get(name, 0) + frequency

		elif isinstance(entry, Recipe):
			for item in entry.required_items:
				resolve_item(item.name, frequency * item.quantity)

	try: # for ValueError case
		recipe = cookbook[name]
		for fin_item in recipe.required_items:
			resolve_item(fin_item.name, fin_item.quantity)
	except ValueError:
		return 'Recipe contains unknown items', 400
	
	ingredients_list = [
		{"name": item_name, "quantity": item_quantity}
		for item_name, item_quantity in final_ingredients_list.items()
	]

	return jsonify({
		"name": name,
		"cookTime": total_cook_time,
		"ingredients": ingredients_list
	}), 200

# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
	app.run(debug=True, port=8080)

File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";

// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook: Record<string, recipe | ingredient> = {};

// Task 1 helper (don't touch)
app.post("/parse", (req:Request, res:Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  } 
  res.json({ msg: parsed_string });
  return;
  
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {
  recipeName = recipeName.replace(/[_-]/g, " ");
  // regex
  recipeName = recipeName.replace(/[^a-zA-Z\s]/g, ""); // a-z, A-Z, or space (\s) with empty string

	const words: string[] = recipeName.trim().split(/\s+/);

	if (words == null) {
    return null;
	
  }

  return words.map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase()).join(" ");
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req:Request, res:Response) => {
  const data = req.body;

	const type: string = data.type;
	const name: string = data.name;
  let new_entry: recipe | ingredient;

	if (name in cookbook) {
    return res.status(400).send("Error: entry already exists! (entry names must be unique)");
  }
	
	if (type == "recipe") {
    const visited_items = new Set<string>();
    const approved_items: requiredItem[] = [];

    const items: any[] = data.requiredItems ?? [];

		for (const item of items) {
      const item_name: string = item.name;
      const item_quantity: number = item.quantity;
      if (visited_items.has(item_name)) {
        return res.status(400).send("Error: duplicated required item in recipe!");
      }

      visited_items.add(item_name);
      approved_items.push({
        name: item_name,
        quantity: item_quantity
      });
    }

    new_entry = {
      name: name,
      type: "recipe",
      requiredItems: approved_items
    };
  } else if (type == "ingredient") {
    const cook_time: number = data.cookTime;
		
		if (cook_time < 0) {
      return res.status(400).send("Error: invalid cook time! (cook time >= 0)");
    }
		
		new_entry = {
      name: name,
      type: "ingredient",
      cookTime: cook_time
    };

  } else {
    return res.status(400).send("Error: invalid type! (not recipe or ingredient)");
  }
	
	cookbook[name] = new_entry;
	
	return res.status(200).json({});

});

// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req:Request, res:Response) => {
  const name: string = req.query.name as string;
  if (!(name in cookbook) || cookbook[name].type !== "recipe") {
    return res.status(400).send("Error: name not found or is not a recipe!");
  }

  let total_cook_time: number = 0;
  const final_ingredients_list: Record<string, number> = {};

	function resolve_item(name: string, frequency: number): void {
    if (!(name in cookbook)) {
      throw new Error("Missing item in cookbook");
    }

    const entry = cookbook[name];

    if (entry.type === "ingredient") {
      const ingredient = entry as ingredient;
      total_cook_time += ingredient.cookTime * frequency;
      final_ingredients_list[name] = (final_ingredients_list[name] ?? 0) + frequency;
    } else if (entry.type === "recipe") {
      const recipe = entry as recipe;
      for (const item of recipe.requiredItems) {
        resolve_item(item.name, frequency * item.quantity);
      }
    }
  }

  try {
    const recipe = cookbook[name] as recipe;

    for (const fin_item of recipe.requiredItems) {
      resolve_item(fin_item.name, fin_item.quantity);
    }
  } catch (err) {
    return res.status(400).send("Recipe contains unknown items");
  }

  const ingredients_list = Object.entries(final_ingredients_list).map(
    ([item_name, item_quantity]) => ({
    name: item_name,
    quantity: item_quantity
    })
  );

  return res.status(200).json({
    name: name,
    cookTime: total_cook_time,
    ingredients: ingredients_list,
  });
});

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});
